zookeeper-server/src/main/java/org/apache/zookeeper/ZooKeeper.java

```diff
@@ -3142,74 +3148,53 @@ public void removeAllWatches(String path, WatcherType watcherType, boolean local
      * @throws KeeperException If the server signals an error with a non-zero
      *  error code.
      */
-    public void addPersistentWatch(String basePath, Watcher watcher, boolean recursive)
+    public void addWatch(String basePath, Watcher watcher, AddWatchMode mode)
             throws KeeperException, InterruptedException {
         PathUtils.validatePath(basePath);
         String serverPath = prependChroot(basePath);
 
         RequestHeader h = new RequestHeader();
-        h.setType(ZooDefs.OpCode.addPersistentWatch);
-        AddPersistentWatcherRequest request = new AddPersistentWatcherRequest(serverPath, recursive);
+        h.setType(ZooDefs.OpCode.addWatch);
+        AddWatchRequest request = new AddWatchRequest(serverPath, mode.getMode());
         ReplyHeader r = cnxn.submitRequest(h, request, new ErrorResponse(),
-                new PersistentWatchRegistration(watcher, basePath, recursive));
+                new AddWatchRegistration(watcher, basePath, mode));
         if (r.getErr() != 0) {
             throw KeeperException.create(KeeperException.Code.get(r.getErr()),
                     basePath);
         }
     }
 
     /**
-     * Async version of {@link #addPersistentWatch(String, Watcher, boolean)} (see it for details)
+     * Async version of {@link #addWatch(String, Watcher, AddWatchMode)} (see it for details)
      *
-     * @param basePath the top path that the watcher applies to
+     * @param basePath the path that the watcher applies to
      * @param watcher the watcher
-     * @param recursive if true applies not only to the registered path but all child paths recursively including
-     *                  any child nodes added in the future
+     * @param mode type of watcher to add
      * @param cb a handler for the callback
      * @param ctx context to be provided to the callback
      * @throws IllegalArgumentException if an invalid path is specified
      */
-    public void addPersistentWatch(String basePath, Watcher watcher, boolean recursive,
-                                   VoidCallback cb, Object ctx) {
+    public void addWatch(String basePath, Watcher watcher, AddWatchMode mode,
+                         VoidCallback cb, Object ctx) {
         PathUtils.validatePath(basePath);
         String serverPath = prependChroot(basePath);
 
         RequestHeader h = new RequestHeader();
-        h.setType(ZooDefs.OpCode.addPersistentWatch);
-        AddPersistentWatcherRequest request = new AddPersistentWatcherRequest(serverPath, recursive);
+        h.setType(ZooDefs.OpCode.addWatch);
+        AddWatchRequest request = new AddWatchRequest(serverPath, mode.getMode());
         cnxn.queuePacket(h, new ReplyHeader(), request, new ErrorResponse(), cb,
-                basePath, serverPath, ctx, new PersistentWatchRegistration(watcher, basePath, recursive));
+                basePath, serverPath, ctx, new AddWatchRegistration(watcher, basePath, mode));
```

Comment:
From other example: we also need a `AddWatchResponse`, the `new ErrorResponse()` is a little tricky ?

Reply:
Anyone else agree? I don't remember why I did it this way TBH.

Reply:
I did a quick check and `ErrorResponse` doesn't bother me. It seems a convenient response for requests like this. But, other committers should have the final word.